usr/sbin/p11sak/p11sak.c

```diff
@@ -2699,6 +2701,11 @@ static CK_RV list_ckey(CK_SESSION_HANDLE session, CK_SLOT_ID slot,
                 fprintf(stderr, "Object does not have CKA_CLASS attribute (error code 0x%lX: %s)\n",
                         rc, p11_get_ckr(rc));
                 p11_uri_attributes_free(uri);
+                p11_uri_free(uri);
+                free(label);
+                label = NULL;
+                free(keytype);
+                keytype = NULL;
```

Comment:
I don't like this code duplication. Maybe we can use a label at the end of the while loop to do all the cleanup. 

Or we move the content of the while-loop to a function, which only take care about a single key. This would make the cleanup-stuff much easier. The more I think about it, I personally would prefer this single-key-function solution.

Reply:
How about this:

```
 usr/sbin/p11sak/p11sak.c | 37 +++++++++++--------------------------
 1 file changed, 11 insertions(+), 26 deletions(-)

diff --git a/usr/sbin/p11sak/p11sak.c b/usr/sbin/p11sak/p11sak.c
index 5789354c..91343e07 100644
--- a/usr/sbin/p11sak/p11sak.c
+++ b/usr/sbin/p11sak/p11sak.c
@@ -2615,7 +2615,7 @@ static CK_RV list_ckey(CK_SESSION_HANDLE session, CK_SLOT_ID slot,
     CK_INFO info;
     CK_SLOT_INFO slot_info;
     CK_TOKEN_INFO token_info;
-    struct p11_uri *uri;
+    struct p11_uri *uri = NULL;
 
     rc = tok_key_list_init(session, kt, label);
     if (rc != CKR_OK) {
@@ -2653,6 +2653,7 @@ static CK_RV list_ckey(CK_SESSION_HANDLE session, CK_SLOT_ID slot,
         goto done;
     }
 
+    label = NULL;
     while (1) {
         rc = funcs->C_FindObjects(session, &hkey, 1, &count);
         if (rc != CKR_OK) {
@@ -2670,7 +2671,7 @@ static CK_RV list_ckey(CK_SESSION_HANDLE session, CK_SLOT_ID slot,
                 fprintf(stderr,
                         "Retrieval of key type failed (error code 0x%lX: %s)\n",
                         rc, p11_get_ckr(rc));
-            continue;
+            goto cont;
         }
 
         rc = tok_key_get_label_attr(session, hkey, &label);
@@ -2700,13 +2701,7 @@ static CK_RV list_ckey(CK_SESSION_HANDLE session, CK_SLOT_ID slot,
             if (rc != CKR_OK) {
                 fprintf(stderr, "Object does not have CKA_CLASS attribute (error code 0x%lX: %s)\n",
                         rc, p11_get_ckr(rc));
-                p11_uri_attributes_free(uri);
-                p11_uri_free(uri);
-                free(label);
-                label = NULL;
-                free(keytype);
-                keytype = NULL;
-                continue;
+                goto cont;
             }
         }
 
@@ -2715,13 +2710,7 @@ static CK_RV list_ckey(CK_SESSION_HANDLE session, CK_SLOT_ID slot,
             if (rc != CKR_OK) {
                 fprintf(stderr, "Object does not have CKA_ID attribute (error code 0x%lX: %s)\n",
                         rc, p11_get_ckr(rc));
-                p11_uri_attributes_free(uri);
-                p11_uri_free(uri);
-                free(label);
-                label = NULL;
-                free(keytype);
-                keytype = NULL;
-                continue;
+                goto cont;
             }
         }
 
@@ -2730,13 +2719,7 @@ static CK_RV list_ckey(CK_SESSION_HANDLE session, CK_SLOT_ID slot,
             if (rc != CKR_OK) {
                 fprintf(stderr, "Object does not have CKA_LABEL attribute (error code 0x%lX: %s)\n",
                         rc, p11_get_ckr(rc));
-                p11_uri_attributes_free(uri);
-                p11_uri_free(uri);
-                free(label);
-                label = NULL;
-                free(keytype);
-                keytype = NULL;
-                continue;
+                goto cont;
             }
         }
 
@@ -2751,9 +2734,6 @@ static CK_RV list_ckey(CK_SESSION_HANDLE session, CK_SLOT_ID slot,
             printf("\n      Attributes:\n");
         }
 
-        p11_uri_attributes_free(uri);
-        p11_uri_free(uri);
-
         switch (keyclass) {
         case CKO_SECRET_KEY:
             rc = sec_key_print_attributes(session, hkey, long_print);
@@ -2796,6 +2776,11 @@ static CK_RV list_ckey(CK_SESSION_HANDLE session, CK_SLOT_ID slot,
                 printf(" %*s | ", CELL_SIZE, keytype);
             printf("\"%s\"\n", label);
         }
+
+cont:
+        p11_uri_attributes_free(uri);
+        p11_uri_free(uri);
+        uri = NULL;
         free(label);
         label = NULL;
         free(keytype);

```

Reply:
Much better. I leave it up to you to stay with this solution or give the single-key-function a try (which I personally still prefer).